### PR TITLE
User can register for an account CLIENT

### PR DESCRIPTION
**Pull request by ViktorHek**
[User can register for an account CLIENT](https://www.pivotaltracker.com/story/show/176470274)
### User Story
```
As an User,
In order to subscribe,
We would like to authenticate myself.
```
## Changes proposed in this pull request:
* Creating a test for visitors can register to read articles
* Creating RegistrationForm using Modal
* Adding logic to hide Sign Up button when registration was successful
* Implementing a function to make registration form displayed when a visitor clicks an article

## Packages/Gems added
* j-tockauth

## Screenshots (Client)
Screenshot 2021-01-13 at 20 22 11

### CODE REVIEW

File: cypress/integration/userCanRegisterToBuySubscription.feature.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            errors: {
+              full_messages: ["Invalid login credentials. Please try again."],
+            },
+          },
+          success: false,
+        },
+      });
+      cy.visit("/");
+    });
+    it("with invalid credentials", () => {
```
Comment (ViktorHek) [inline cypress/integration/userCanRegisterToBuySubscription.feature.js]: add a sad path for "this email is already taken"

File: cypress/integration/userCanRegisterToBuySubscription.feature.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        response: "fixture:user_can_register.json",
+      });
+      cy.visit("/");
+    });
+    it("with valid credentials", () => {
+      cy.get("[data-cy='signup-button']").click();
+      cy.get("[data-cy='input-email']").type("registered@user.com");
+      cy.get("[data-cy='input-password']").type("password");
+      cy.get("[data-cy='input-password-confirmation']").type("password");
+      cy.get("[data-cy='submit-btn']").click();
```
Comment (ViktorHek) [inline cypress/integration/userCanRegisterToBuySubscription.feature.js]: ad this inside ```cy.get("[data-cy='registration-form']").within(() => {}```

File: cypress/integration/userCanRegisterToBuySubscription.feature.js (hunk 1 above)
Comment (KyunginNa) [reply cypress/integration/userCanRegisterToBuySubscription.feature.js]: We thought it'd be okay not to test all the sad paths but just one, because it is built-in function in rails and we all know it works well...

File: cypress/integration/userCanRegisterToBuySubscription.feature.js (hunk 2 above)
Comment (KyunginNa) [reply cypress/integration/userCanRegisterToBuySubscription.feature.js]: Added!

File: cypress/integration/userCanRegisterToBuySubscription.feature.js (hunk 1 above)
Comment (KyunginNa) [reply cypress/integration/userCanRegisterToBuySubscription.feature.js]: I'll add it! :)

File: src/components/RegistrationForm.jsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+const RegistrationForm = () => {
+  const { registerErrorMessage, modalOpen } = useSelector(state => state);
+  const dispatch = useDispatch();
+
+  return (
+    <>
+      <Modal
+        closeIcon
+        onClose={() => dispatch({ type: "CLOSE_MODAL" })}
+        onOpen={() => dispatch({ type: "OPEN_MODAL" })}
```
Comment (ViktorHek) [inline src/components/RegistrationForm.jsx]: not really a problem a.t.m but when we add the `subscriber modal` we might want to change this name "CLOSE_REGISTATION" and "OPEN_REGISTRATION"

Comment (KyunginNa) [reply src/components/RegistrationForm.jsx]: Good thinking!⭐️ Changed them to OPEN_REGISTRATION_FORM and CLOSE_REGISTRATION_FORM

File: cypress/integration/userCanReadASpecificArticle.feature.js [hunk 4]
```diff
@@ -37,7 +56,7 @@ describe("User can", () => {
     });
   });
 
-  context("not see an article if improper params are provided", () => {
+  describe("not see an article if improper params are provided", () => {
```
Comment (emtalen) [inline cypress/integration/userCanReadASpecificArticle.feature.js:59]: why did you change this from `context` to `describe`?

File: cypress/integration/userCanReadASpecificArticle.feature.js [hunk 5]
```diff
@@ -58,4 +94,12 @@ describe("User can", () => {
       );
     });
   });
+
+  describe("NOT read article when they're not registered", () => {
```
Comment (emtalen) [inline cypress/integration/userCanReadASpecificArticle.feature.js:98]: Make sure that this is restricted in your API as well. So not anything you need to change in this PR but take a note on updating your API

File: src/modules/getArticles.js [hunk 6]
```diff
@@ -13,12 +13,19 @@ const getArticles = {
...
+        const response = await axios.get(`/articles/${articleId}`);
+        dispatch({ type: "VIEW_ARTICLE", payload: response.data.article });
+      } catch (error) {
+        dispatch({
+          type: "ERROR_MESSAGE",
+          payload: error.response.data.message,
+        });
+      }
+    } else {
+      dispatch({ type: "OPEN_REGISTRATION_FORM" });
```
Comment (emtalen) [inline src/modules/getArticles.js:28]: nice solution for this ✨

File: cypress/integration/userCanReadASpecificArticle.feature.js (hunk 4 above)
Comment (KyunginNa) [reply cypress/integration/userCanReadASpecificArticle.feature.js:59]: We have `describe` in other tests, so it was just for consistency!

**Review by emtalen — APPROVED** [review]
